File: src/Hartonomous.Core/Compute/Ingestion/Svd.cs
using System;
using Hartonomous.Core.Compute.Internal;

namespace Hartonomous.Core.Compute.Ingestion;

/// <summary>
/// Thin singular value decomposition (MKL dgesdd, divide-and-conquer).
/// For A ∈ R^(m×n) row-major with k = min(m, n): A = U · diag(S) · V^T,
/// U ∈ R^(m×k) row-major, S ∈ R^k descending, V^T ∈ R^(k×n) row-major.
/// Deterministic under MKL CBWR=AUTO,STRICT. A is preserved.
/// Caller owns all output buffers.
/// </summary>
public static class Svd
{
    /// <summary>Minimum singular value count for a given shape (= min(m, n)).</summary>
    public static long MinDim(long m, long n) => m < n ? m : n;

    /// <summary>
    /// Compute the thin SVD of a row-major m×n f64 matrix.
    /// </summary>
    /// <param name="m">Row count. Must be &gt; 0.</param>
    /// <param name="n">Column count. Must be &gt; 0.</param>
    /// <param name="a">Input matrix, row-major, length &gt;= m·n.</param>
    /// <param name="u">Output left singular vectors, row-major, length &gt;= m·k.</param>
    /// <param name="s">Output singular values descending, length &gt;= k.</param>
    /// <param name="vt">Output right singular vectors transposed, row-major, length &gt;= k·n.</param>
    public static void F64(
        long m, long n,
        ReadOnlySpan<double> a,
        Span<double> u,
        Span<double> s,
        Span<double> vt)
    {
        if (m <= 0 || n <= 0)
        {
            throw new ComputeArgumentException($"svd_f64: invalid shape m={m}, n={n}");
        }
        long k = MinDim(m, n);
        long aLen = checked(m * n);
        long uLen = checked(m * k);
        long vtLen = checked(k * n);
        if (a.Length < aLen)
        {
            throw new ComputeArgumentException($"svd_f64: input buffer too small ({a.Length} < {aLen})");
        }
        if (u.Length < uLen)
        {
            throw new ComputeArgumentException($"svd_f64: U buffer too small ({u.Length} < {uLen})");
        }
        if (s.Length < k)
        {
            throw new ComputeArgumentException($"svd_f64: S buffer too small ({s.Length} < {k})");
        }
        if (vt.Length < vtLen)
        {
            throw new ComputeArgumentException($"svd_f64: V^T buffer too small ({vt.Length} < {vtLen})");
        }
        NativeError.ThrowIfError(
            NativeCompute.SvdF64(m, n, a, u, s, vt),
            "svd_f64");
    }
}

File: ext/libhartonomous/src/svd.c
/*
 * svd.c — dense f64 singular value decomposition via MKL dgesdd (divide-
 *          and-conquer). Row-major throughout so callers stay in C/C#
 *          conventions. Computes the thin SVD: for A in R^(m×n) with
 *          k = min(m, n),
 *             A = U · diag(S) · V^T,
 *          U  ∈ R^(m×k), S ∈ R^k (descending), V^T ∈ R^(k×n).
 *
 * Determinism. MKL dgesdd is deterministic under CBWR=AUTO,STRICT. Reduction
 * order is fixed, no PRNG is involved, and LAPACK's row-major wrapper is a
 * pure argument marshalling layer — same inputs produce bit-identical
 * outputs across repeated runs on the same ISA.
 *
 * This is the only SVD used by the substrate (Procrustes alignment,
 * attention-pattern right-singular vectors, tensor-energy spectra, feature
 * clustering preprocessing). No randomized / truncated / quantized variants
 * are offered. Law #6 forbids approximation; if the substrate needs a
 * sparse Lanczos SVD later, it will be a separately-named primitive
 * against an explicit sparse matrix.
 */

#include "hartonomous.h"

#include <stddef.h>
#include <stdint.h>

#include <mkl.h>
#include <mkl_lapacke.h>

int hartonomous_svd_f64(
    int64_t m, int64_t n,
    const double* a,
    double* u,
    double* s,
    double* vt
) {
    if (a == NULL || u == NULL || s == NULL || vt == NULL) {
        return -1;
    }
    if (m <= 0 || n <= 0) {
        return -2;
    }

    const int64_t k = (m < n) ? m : n;

    /* dgesdd overwrites the input matrix when jobz='O'; with jobz='S' it
     * leaves A intact and writes the thin U and V^T. We want A preserved,
     * so copy into a scratch buffer sized m·n. */
    double* a_copy = (double*)mkl_malloc((size_t)m * (size_t)n * sizeof(double), 64);
    if (a_copy == NULL) {
        return -9;
    }
    /* Plain copy — memcpy is fine; MKL's cblas_dcopy could be used but adds
     * no value at this size. */
    for (int64_t i = 0; i < m * n; ++i) {
        a_copy[i] = a[i];
    }

    /* LAPACKE row-major:
     *   jobz = 'S'  → thin U (m × k) and thin V^T (k × n)
     *   lda = n     (row-major stride of A)
     *   ldu = k     (row-major stride of U when thin)
     *   ldvt = n    (row-major stride of V^T)
     */
    lapack_int info = LAPACKE_dgesdd(
        LAPACK_ROW_MAJOR,
        'S',
        (lapack_int)m,
        (lapack_int)n,
        a_copy,
        (lapack_int)n,
        s,
        u,
        (lapack_int)k,
        vt,
        (lapack_int)n
    );

    mkl_free(a_copy);

    if (info > 0) {
        /* DBDSDC did not converge. Treat identically to Lanczos non-convergence
         * (error code -6) so the facade can surface a uniform diagnostic. */
        return -6;
    }
    if (info < 0) {
        /* Invalid argument — the argument marshalling above should prevent
         * this, so treat as a shape error for the caller's contract. */
        return -2;
    }

    (void)k;
    return 0;
}
